sycl/doc/extensions/experimental/sycl_ext_codeplay_kernel_fusion.asciidoc

```diff
@@ -0,0 +1,678 @@
+`complete_fusion` or `cancel_fusion` are called.
+
+If the wrapped `queue` is already in fusion mode, the function throws an
+`exception` with `errc::invalid` error code.
+
+|`event complete_fusion(const property_list &)` 
+
+|Complete the fusion: If the
+runtime decides to perform fusion, it will JIT-compile a fused kernel from all
+kernels submitted to the wrapped `queue` since the last call to `start_fusion`
+and submit the fused kernel for execution. Inside the fused kernel, the
+per-work-item effects are executed in the same order as the kernels were
+initially submitted, adding group barriers between each of them by default.
+Otherwise, the individuals kernels will be passed to the scheduler and executed
+in the same order as they were initially submitted. Constraints on when fusion
+is possible and criteria for the implementation to perform fusion are
+implementation-defined. Calling `fusion_wrapper::complete_fusion` does therefore
+not guarantee that the kernels will be fused.
+
+The call is asynchronous, i.e., it may return after fusion (JIT-compilation) is
+done, but before execution of the fused kernel is completed. The returned event
+allows to synchronize with the execution of the fused kernel.
+
+At call completion the wrapped `queue` is no longer in fusion mode, until the
+next `start_fusion`.
+
+|`void cancel_fusion()` 
+
+|Cancel the fusion and submit all kernels submitted to the wrapped `queue` since
+the last `start_fusion()` for immediate execution **without** fusion. The
+kernels are submitted in the same order as they were initially submitted to the
+queue.
+
+This operation is asynchronous, i.e., it may return after the kernels have been
+added to the scheduler, but before any of the previously submitted kernel starts
+or completes execution.
+
+At call completion the wrapped `queue` is no longer in fusion mode, until the next
+`start_fusion`.
+
+|`bool is_in_fusion_mode() const`
+|Returns true if the wrapped SYCL `queue` is currently in fusion mode.
+
+|===
+
+==== Properties
+
+Next to the new API functions and classes described above, this extension also
+adds new properties that are described in Table 3.
+
+Table 3. New properties for kernel fusion.
+
+|===
+|Property|Description
+
+|`sycl::ext::codeplay::property::queue::enable_fusion`
+|This property enables kernel fusion for the queue. If a `fusion_wrapper` object
+is constructed on a queue without this property, an `exception` with `errc::invalid`
+error code is thrown.
+
+|`sycl::ext::codeplay::property::no_barriers`
+
+|If the property list passed to `fusion_wrapper::complete_fusion()` contains this
+property, no barriers are introduced between kernels in the fused kernel.
+
+|`sycl::ext::codeplay::property::promote_local`
+|This property gives a hint to the JIT compiler to try to internalize a given
+ argument via promotion to local memory (see local and private internalization
+ in <<internalization>>).
```

Comment:
This property (and the next one) are not hints.  They are assertions by the programmer that the kernels use buffer elements in a certain way.  The wording should state this more clearly.  For example, this might be a description of `promote_private`:

> This property is an assertion by the application that each element in the buffer is accessed by no more than one work-item in any kernel in the fusion set.  Implementations may treat this as a hint to promote the buffer elements to private memory.

It's important for the descriptions here to clearly state the condition that must be true in order to safely set these properties.  The formal specification of these conditions should be here in the description of the properties, rather than relying on the "Overview" section.

Reply:
Agreed, thanks for the feedback. We have updated the description of `promote_local` and `promote_private` with a text similar to your proposal.